### DISCUSSION TITLE
Change the Database model

### DISCUSSION THREAD

**Discussion by entchen66** [kkrypt0nn/Python-Discord-Bot-Template#78] [Ideas]
Hi there,

As a good database option, I would suggest changing the model so that each cog has its own database for users, guilds, and global settings. This can be achieved by initializing the settings in the `__init__` method of each cog, along with a random ID (to prevent duplications when cogs have the same name).

Furthermore, settings can be initialized in the `__init__` function, such as global_settings, guild_settings, member_settings, channel_settings, role_settings, etc. These settings can then be retrieved in the cog while using the initialized settings object.

Available methods for example are:

*    `await self.settings.guild(guild_ID).parameter_name()` - get the config
*    `await self.settings.guild(guild_ID).parameter_name().set(value)` - set the config

The value can be of type dict, list, string, or int and should be parsed accordingly.

All of this could be implemented in a `db_manager`.

I would be happy to help you more with this implementation if you decide to work on it.

Best regards,
entchen.

**Comment by Nyvil** [comment]
Hey there,

similar to your last suggestion, reworking the code**base** to adjust the database model as you suggest is a bit out of scope for what this **template** is supposed to do. 

This template is supposed to give new developers a head start on their journey as a programmer, not throw them into the cold water with a code level that basically makes it impossible to understand what's going on as a new developer.

Best Regards,
Nyvil

**Reply by entchen66** [reply]
Thank you for your response, Nyvil. I understand that the primary goal of the template is to provide a starting point for new developers to learn from. However, I believe that a good database concept is important for several reasons. A well-designed database allows for better organization of data, making it easier to find, retrieve, and manipulate the information needed. It also allows for easy scalability, enabling the addition of new data and users without compromising performance. Additionally, a good database design ensures data integrity, ensures better security, improves the overall performance of the application and improves the reusability of the code and ease of modification. While adding a good database concept might be more advanced, it's a crucial step in bot development that ensures that a bot can handle the real-world requirements such as the large number of users and data. Additionally, It will be beneficial for future development career of the developers to learn and implement good database concepts early. Thank you again for the template and your consideration.

**Reply by Nyvil** [reply]
Hey entchen,

thank you for your response, and while I agree with some of your points, providing a database management system is prone to errors, and it simply requires more to get started. Example given:

- a database server (-> dedicated server like a root or vps)
- redundancy in case of data corruption -> multiple database servers to ensure data integrity and safety, which, when poorly handled, increases the latency, you'd need a worker to keep the data across all databases coherent
- a system like k8s or docker to ensure scalability and sharding databases 

In addition, while we're keen to help new developers get started on their journey as such, taking away every learning opportunity by providing a ready-to-go solution doesn't help us, nor them. 

I once again thank you for your time and consideration for our project.

Best Regards,
Alexandros
